### ISSUE TITLE
Support limited substitution syntax in env var values

### ISSUE THREAD

**Issue by pmorie** [GoogleCloudPlatform/kubernetes#7739]
I think we should support a limited substitution syntax for env var values.  The specifics:

1.  Should support `$VAR and ${VAR}`
2.  Resolution should happen after runtime-resolved variables are resolved (ie, fieldRef for EnvVarSource)
3.  Should only support kubernetes vars, ie:
    1.  Service env vars
    2.  Env vars in the container's env

@smarterclayton @thockin

**Comment by smarterclayton** [comment]
Context is that docker is not doing this in 1.6 for env, it's almost impossible to actually build URLs to kube services using the env var mechanism, and command + args is horribly ugly and I die a little time everytime I make someone use it.

Here's a relatively real world example of a server running in a namespace that needs to publish its public URL (so that it can provide it for others to access):

https://github.com/smarterclayton/origin/blob/31a74c16259c7f576417990882cbc8696d3b50b6/examples/gitserver/template.yaml

```
          env:
          - name: PUBLIC_URL
            value: "http://gitserver.myproject.local:8080"
```

I can't use DNS or someone can only run this into one namespace (whichever one I hardcode).  I can't use service ports because I can't expand the service env var in the local step.  Ideally I'd be able to do:

```
          env:
          - name: PUBLIC_URL
            value: "http://${GITSERVER_SERVICE_HOST}:${GITSERVER_SERVICE_PORT}"
```

or

```
          env:
          - 
          - name: PUBLIC_URL
            value: "http://gitserver.${POD_NAMESPACE}.local:${SERVICE_PORT}"
```

**Comment by timothysc** [comment]
@thockin, @smarterclayton - 
Isn't this kind of a 1st step into configuration language space here?

**Comment by smarterclayton** [comment]
Technically having ENV vars is probably that step.  This is the next logical step.  I'm happy to draw this line hard, because we already have tools for describing the structural relationships (selectors, objects, etc).  I'd argue more about this is purely to bring workloads into Kube and stop people from creating wrapper layers.  But that's a line in my head, and I'm sure everyone has said that before.

----- Original Message -----

**Comment by bcbroussard** [comment]
I agree with @smarterclayton and @pmorie that finding an easy solution for automatic ENV variable resolution is something good for v1. I've worked with a number of clients who are just starting out with k8s and almost every one of them has hit an issue around this. 

Sometimes it involves confusion around the k8s command/args to docker cmd/entrypoint mappings, but that is another topic.

IMHO, this is one of those little things that can make working with k8s smooth and enjoyable.

**Comment by thockin** [comment]
It's also muddled by Docker's bizarre "sometimes it is a shell, sometimes
not" semantics.  I agree it is pretty obvious, the question I have is
mostly timeframe.

**Comment by timothysc** [comment]
fwiw I was planning on looking into https://github.com/hashicorp/hcl & https://www.terraform.io/docs/configuration/syntax.html in the post 1.0 timeframe to address some other orthogonal issues, but they related to these set of problems. 

However,  I think I'm in agreement with @thockin on timeframe, because it would be difficult if we put something out-there only to deprecate it shortly after.

**Comment by pmorie** [comment]
> It's also muddled by Docker's bizarre "sometimes it is a shell, sometimes not"
semantics.

IMO this is one of the worst pieces of the docker API.

**Comment by pmorie** [comment]
Resolved by #7936 and #8149

**Comment by bgrant0607** [comment]
cc @ghodss @jackgr
